docs: Provide policy on breaking changes & deprecations in KEDA (#70)
kedacore/governance: DEPRECATIONS.md

# Deprecations & Breaking Changes in KEDA

This document describes how KEDA handles deprecations and breaking changes and should provide clarity to KEDA end-users on what to expect when upgrading to a new version of KEDA.

If you are interested in the current open deprecations, we recommend checking our [GitHub Discussions](https://github.com/kedacore/keda/discussions/categories/deprecations) or the open [breaking changes](https://github.com/kedacore/keda/issues?q=is%3Aopen+label%3Abreaking-change+sort%3Aupdated-desc).

## Our versioning strategy

KEDA follows the deprecation strategy of Kubernetes and is allowed to make breaking changes in minor versions after a deprecation period.

> The KEDA runtime, which consists of the operator and metric server, will provide support on the deprecated features for 4 releases/1 year after the deprecation was announced. In the following release, the feature will be removed.

Author: ```suggestion
The KEDA runtime, which consists of the operator and metric server, will provide support on the deprecated features for 4 releases/1 year after the deprecation was announced. In the following release, the feature will be removed.

In some scenarios, a graceful phase-out approach can be used by using feature flags. This allows end-users to keep using the deprecated functionality if they need to.

The phase-out with feature flags would be as follow:
1. A new feature flag is introduced at time of deprecation that enables the deprecated functionality to stay available. If end-users do not depend on it, they can already disable it to avoid taking new dependency on it.
2. In release N+2, the feature flag is turned off by default so that the deprecated functionality is no longer available as a scream test. However, if end-users still rely on it they can enable the functionality again.
3. N+4 releases removes support for the deprecated feature and it will no longer be available
```

Reviewer: I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable then. We are already giving a year for the migration

Reviewer 2: -1 we should try to keep things simple, we should not try to over engineer this. I'd be also for reducing the deprecation from 4 releases to 2 :)

Reviewer: I'd reduce it too, but we decided 4 xD
My question is, does make sense a feature flag for removing features? Isn't the required effort quite similar to just upgrade/downgrade KEDA itself? I mean, changing the yaml to enable/disable the flag looks like just change the version.

Author: > I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable them (apart from the needed code, ofc). We are already giving a year for the migration

This is to get end-users off of it faster instead of only after 1 year.

> I'd reduce it too, but we decided 4 xD

Exactly, why did we decide to 4 if we change it again :)

> My question is, does make sense a feature flag for removing features? Isn't the required effort quite similar to just upgrade/downgrade KEDA itself? I mean, changing the yaml to enable/disable the flag looks like just change the version.

No because if we don't have flags we have to have full support for 4 versions long while with feature flags we could start "hiding" things so that less people use it by default.

Author: If nobody wants this then we can leave this out but this is exactly what we are considering at the moment :)

Reviewer: > Exactly, why did we decide to 4 if we change it again :)

I don't want to change it, even I'd do it xD (We discussed this, and I respect our decision)

> No because if we don't have flags we have to have full support for 4 versions long while with feature flags we could start "hiding" things so that less people use it by default.

This is a user decision, if end-user decides to use a deprecated feature as default, it's totally their problem IMO, we cannot do anything there as they can enable the flag. We have to keep the deprecated code for all the deprecation period and we are not babysitters, our users are adults, they can read a deprecation message during 1 year

Author: > This is a user decision, if end-user decides to use a deprecated feature as default, it's totally their problem IMO, we cannot do anything there as they can enable the flag. We have to keep the deprecated code for all the deprecation period and we are not babysitters, our users are adults, they can read a deprecation message during 1 year

You are assuming they are aware of deprecations, though, which is not an assumption that I would make. Not everyone is closely watching KEDA.

When building products it's best to help customers migrate, or they will not do it until it is (almost) too late

Reviewer: > You are assuming they are aware of deprecations, though, which is not an assumption that I would make. Not everyone is closely watching KEDA.

We are going to log them, if they aren't aware after a year of warnings, they will try to update KEDA, facing the "breaking change", I mean, they can be aware if they want, and they won't be aware if they don't. We are going to publish deprecation on several channels, including the pod logs.

> When building products it's best to help customers migrate, or they will not do it until it is (almost) too late

Agreed, but I prefer to keep it simple, as this is not like k8s upgrade that upgrades a lot of components, this is just rolling back to the previous version which means a few clicks

Reviewer 2: > > I'm not really sure about this, this will be more work documenting the feature flags and how to enable/disable them (apart from the needed code, ofc). We are already giving a year for the migration
> 
> This is to get end-users off of it faster instead of only after 1 year.
> 
> > I'd reduce it too, but we decided 4 xD
> 
> Exactly, why did we decide to 4 if we change it again :)

well, we said that we will revisit this version of the proposal 😄  I am okay with keeping 4 releases, but after I gave it a thought, it seems to be quite long for me.

Author: Let's lave it out

> Depending on the area of the change, KEDA might not wait 4 releases to introduce a breaking change because one of its dependencies such as Kubernetes or scaler dependencies. However, these are the only exceptions to the rule.

Author: This is a proposal by @zroubalik 
```suggestion
Depending on the area of the change, KEDA might not wait 2 releases to introduce a breaking change because one of its dependencies such as Kubernetes or scaler dependencies. However, these are the only exceptions to the rule.
```

Any preferences @jeffhollan @JorTurFer @ahmelsayed? This means supporting something for 6mo / 2 releases.

(however, we might want to explicitly say "6mo (equivalent of 2 releases)" in case we decide to start shipping faster)

Reviewer: I can live with 4 (1 year) but I prefer 2 releases (6 months, so this is okay to me

Reviewer 3: Personally, I'd always prefer longer deprecation cycles, but I don't want to deadlock the decision since I'm ok with either. 
though I'm a bit confused by the 6 month thing, seems Jorge you suggesting we guarantee a minimum of 6 months notice, while Tom you are saying 2 releases regardless of release cadence? or did I misunderstand that?

Reviewer 3: I see, is that what you meant Tom? That sounds good to me

Author: Yes, that is correct. Updated in [`3dd6427` (#70)](https://github.com/kedacore/governance/pull/70/commits/3dd642773f590d22d5012de027e9d74ef86ce10f) which is purely based on time and uses releases as a reference

### Custom Resource Definitions (CRDs)

Our custom resource definitions (CRDs) have a separate versioning scheme than the KEDA runtime given they have an `apiVersion`. For this, we follow the [official Kubernetes API versioning policy](https://kubernetes.io/docs/reference/using-api/#api-versioning).

> KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.

Reviewer 4: ```suggestion
KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced. Any of those breaks would be coupled with a KEDA versioned release, with deprecations of previous `apiVersion` held for major version releases.
```

LMK if this is accurate (esp. the waiting to deprecate old api versions with major versions

Author: This is actually not correct, I think as the api versions will be removed in the next minor version.

Thoughts @zroubalik @JorTurFer ?

Reviewer 4: Isn't deprecating a previously supported configuration a breaking change?

Reviewer 4: Oh wait I'm sorry this is about KEDA keeping up with K8s APIs? There it makes sense we'd keep up with their latest versions which are gated behind semver

# Deprecations & Breaking Changes in KEDA

This document describes how KEDA handles deprecations and breaking changes and should provide clarity to KEDA end-users on what to expect when upgrading to a new version of KEDA.

If you are interested in the current open deprecations, we recommend checking our [GitHub Discussions](https://github.com/kedacore/keda/discussions/categories/deprecations) or the open breaking changes](https://github.com/kedacore/keda/issues?q=is%3Aopen+label%3Abreaking-change+sort%3Aupdated-desc).

## Our versioning strategy

KEDA follows the [Semantic Versioning](https://semver.org/) strategy and can only introduce breaking changes in new major versions.

However, depending on the nature of the change, we may introduce deprecations in existing minor versions. These changes always must be backwards-compatible.

> Depending on the area of the change, we might not need to wait for the next major version of KEDA to introduce a breaking change.

Reviewer 2: How should I read this?

> These changes always must be backwards-compatible.

Does that mean, that we will support a previous version of the changed stuff forever?

Author: If we make a change in a minor version (not CRD based) we must be backwards compatible until next major version.

So if we decide to change something to our Prometheus integration, for example, we can but it should not impact existing functionality by default.

Totally fictious scenario to give an indication - If we introduce `endpointUri` instead of `Uri`, we must support uri until 3.x.

Reviewer: I agree that there are some things that we can't delete, but IMO not breaking the compatibility doesn't mean that we will support and add new features to a deprecated thing.
Let's say that we unify prometheus, obviously we should support the old port till a major version, but I wouldn't ensure that the old endpoint will expose new metrics in the future, once a feature is deprecated, I'd keep it as it is till next major, if any user wants the new metrics, they should move to the new endpoint.
WDYT?

Author: I think that's something to be determined on a feature-per-feature basis. If the effort is 0 to add it to both, then why not? It's up to them to still choose if they will migrate or not.

However, the sooner they will the better because we will cut it off at some point anyway.

So long story short - How aggressive do we want to be? I'd say case-by-case by not adding to deprecated things should be the norm for sure.

Let me put that in there that that is how it is, and if we still add something to a deprecated feature then no biggy; but we don't doc it

Author: Updated after our discussion

### KEDA runtime

The KEDA runtime, which consists of the operator and metric server, is not allowed to make breaking changes unless the major version is incremented.

There are no exceptions to this rule.

### Custom Resource Definitions (CRDs)

> Our custom resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.

Reviewer 2: I'd prefer if we link kubernetes policy here. alpha/beta/stable

Author: What about something like this?

```suggestion
Our custom resource definitions (CRDs) have a separate versioning scheme than the KEDA runtime given they have an `apiVersion`. For this, we follow the [official Kubernetes API versioning policy](https://kubernetes.io/docs/reference/using-api/#api-versioning).
```

> Our customer resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.

Reviewer: ```suggestion
Our custom resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.
```

KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.

> However, the KEDA runtime must support all current API versions until the next major version of KEDA.

Reviewer: At this moment, CRDs are in v1alpha1, I'm not saying that we shouldn't avoid breaking changes, but based on that version, we could introduce any breaking change that we want, including any element inside it like triggers or metadata.
If we are going to block breaking changes, I'd promote CRDs to v1 directly

Reviewer: In the same way, I wouldn't support multiple versions till we graduate the api version, alpha or beta are literally versions which allow changes

Author: > At this moment, CRDs are in v1alpha1, I'm not saying that we shouldn't avoid breaking changes, but based on that version, we could introduce any breaking change that we want, including any element inside it like triggers or metadata.

It's not because you can, that you should. That would be a bad customer experience in my opinion.

I think the key take aware with our current situation and this doc is:
1. We should go to v1 soon
2. We should only support 1 stable and one preview API version for our CRD

Author: Once a new stable or preview version is introduced; we maintain it for X versions (ie 2) and kill it afterwards.

Reviewer: Okey but, let's say that we release `scaledobjects.keda.sh/v1` in v2.9.0, and after several releases, e.g in KEDA v2.16 we release `scaledobjects.keda.sh/v2`. Based on this, we will automatically drop the support to `scaledobjects.keda.sh/v1` but that's a breaking change and we can't do it till KEDA v3.

If we try to not doing a v3 release, I think that we should be able to introduce breaking changes with minor versions. In fact, Kubernetes does it, atm k8s is in major v1 and  they introduce breaking changes with minor versions. Personally, I'd go closer with k8s feature lifecycle than semver, because we are creating a software around k8s. 

Other option could be doing a major release per year, for instance (but I think we don't want that).

Author: > Okey but, let's say that we release `scaledobjects.keda.sh/v1` in v2.9.0, and after several releases, e.g in KEDA v2.16 we release `scaledobjects.keda.sh/v2`. Based on this, we will automatically drop the support to `scaledobjects.keda.sh/v1` but that's a breaking change and we can't do it till KEDA v3.

Wrong, that's why the doc explicitly calls out API version and the fact that they have a separate lifecycle from the KEDA runtime so that we are allowed to do this. And for a reason, because they are completely different.

See proposal:

> Depending on the area of the change, we might not need to wait for the next major version of KEDA to introduce a breaking change.
> [...]
> Our custom resource definitions (CRDs) have a seperate versioning from the KEDA runtime given they have an `apiVersion`.
> KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.
>  
> However, the KEDA runtime must support all current API versions until the next major version of KEDA.

So if `scaledobjects.keda.sh/v2` is released, it is going to deprecate/remove (to define) `scaledobjects.keda.sh/v1`. I'd argue that `scaledobjects.keda.sh/v2alpha` or `scaledobjects.keda.sh/v2beta` supersedes `scaledobjects.keda.sh/v2` but the policy needs to be more explicit on this.

My proposal would be:
- A new stable CRD must have a preview CRD that supersedes it
- Once new stable CRD is introduced, the previous stable CRD version is only supported for 2 versions (worst case: 1) so that end-users can migrate
- We only support one preview API version

> Other option could be doing a major release per year, for instance (but I think we don't want that).

This is something I want to avoid.

Does this clarify things @JorTurFer @zroubalik ?

Author: Updated after our discussion

> ### Autoscaling triggers and its metadata

Reviewer: ```suggestion
### Autoscaling triggers and their metadata
```

> However, the KEDA runtime must support all current API versions until the next major version of KEDA.

Reviewer: I'd follow the same approach as CRD content, N+2.
WRT the support of old versions, what does it mean? Will we add new features to both or the old version will enter in "maintenance mode" being provided as it was at the moment that we released the new one?

Until KEDA introduces a [versioning scheme for autoscaling triggers](https://github.com/kedacore/keda/issues/613), it is not allowed to make breaking changes for triggers.

> The only exception to this rule is when a new `apiVersion` is introduced.

Reviewer: IMO we shouldn't maintain more than 1 version, I don't have any problem keeping old versions in the code till next major KEDA versions, but we should define the if we are going to maintain multiple versions or only 1-2.

My vote is for maintaining only the latest version, once we create a new version, the previous one enters automatically in deprecated state in X KEDA versions (ore even more, deprecated versions could be deleted e.g, after 2 KEDA versions). But as I said in other comments, alpha/beta means other thing, doing this we are treating as stable one api which isn't stable yet

Author: I think https://github.com/kedacore/governance/pull/70#discussion_r968071392 also covers this discussion.

Reviewer 2: Agree with @JorTurFer on this.

Reviewer 2: Also, `v1alpha1` `apiVersion` allows us to do any change whenever we want.

Author: @zroubalik see https://github.com/kedacore/governance/pull/70#discussion_r968071392 but it's not because you can that you should.

Reviewer: As we commented in the last standup, I wouldn't keep deprecated fields/behaviours related with CRDs (including the scaler behaviour) for more than X KEDA versions (for me 2). Once we have introduced the scaler versioning, we can discuss how we maintain the each scaler version, but ATM I prefer to maintain them only 2 KEDA versions. 
2 KEDA versions means 6 months approx to address the changes, I think that it's enough time to do it. 
As we discussed, too, I'd add a clear deprecation log saying that the feature will be removed in version XXXX (N+2).

Author: That sounds good to me; but I really think we should respect apiVersions. So are you suggesting that API versions will get bumped after 2 versions? Because solely relying on KEDA version and ignoring the CRD version for this is not correct imo.

Reviewer: I'm not totally sure if I understand your point. I suggest removing every CRD related deprecation 2 version later. If we deprecate spec.XXXX in version 2.9, that deprecated code will be removed in version 2.11. 

I understand your concerns about semantic versioning, but Kubernetes itself doesn't follow it and we are creating a project on top of Kubernetes. E.g: Kubernetes deprecated autoscalingv2/beta2 in v1.23, and they will remove it in v1.26, the major version is 1, and they are removing the whole API, not only a field (K8s does it too in case of simple fields).

In general, I agree that we should follow semver for KEDA itself, but for everything related with CRDs and K8s I wouldn't do it, if we are on top of k8s, we should follow k8s approach

Author: Updated after our discussion

> Until KEDA introduces a [versioning scheme for autoscaling triggers](https://github.com/kedacore/keda/issues/613), it is not allowed to make breaking changes for triggers.

Reviewer: If we finally change the deprecation policy, this line should be modified because we will allow breaking changes with conditions

Author: > If we finally change the deprecation policy

What are you referring to here? This policy or another one that I'm not are of?

Reviewer: I'm referring to this https://github.com/kedacore/governance/pull/70/files#r984894940.
`it is not allowed to make breaking changes for triggers.` sounds too "categorical" because if we support deprecated fields for 2 versions, we will introduce breaking changes. 
In the same way, even we introduce scaler versions, if we are going to maintain each scaler version for 2 versions, we will introduce breaking changes.

Reviewer: Even if we don't remove old scaler version after 2 KEDA versions, we have talked about not maintain a bunch of old versions for each trigger, so eventually we are going to introduce a breaking change related with this because we don't have plans for KEDA v3

Author: Updated after our discussion

This kind of change will be announced in the release notes and enforced in our Helm charts.

## What is considered a breaking change?

A breaking change is a change that is not backwards-compatible and requires end-users to take manual action to migrate or lose existing functionality.

## Avoiding breaking changes

When contributors want to introduce a new approach, they are allowed to do so without breaking changes. This means that the new approach will be available in the next minor version of KEDA but the existing functionality still works.

Every occasion of this must provide warning(s) in the KEDA logs to make customers aware of the change so they can decide whether to migrate to the new approach - Either by using the new functionality in the next minor version or by migrating to the new approach when the next major version was released.

Every deprecation must follow our process described below.

## Introducing new deprecations

When introducing a new deprecation, it must comply with the following rules:

- Every deprecation must be announced in the release notes and will be highlighted in the upcoming release

> - Every deprecation must have a warning in the KEDA logs to create awareness

Reviewer: I'd clarify in the logs message the version that will remove the feature, something like: 
`WARNING: This feature/parameter/whatever has been deprecated in version N and will be removed in version XX.XX, use feature/parameter/whatever instead`
Depending on the type of deprecation (core or CRDs+Scalers) we can say v3.0 or v2.X

Author: Updated after our discussion

## Our versioning strategy

KEDA follows the deprecation strategy of Kubernetes and is allowed to make breaking changes in minor versions after a deprecation period.

The KEDA runtime, which consists of the operator and metric server, will provide support on the deprecated features for 6 months (currently the equivalent of 2 releases) after the deprecation was announced. In the following release, the feature will be removed.

Depending on the area of the change, KEDA might not wait the above mentioned deprecation period to introduce a breaking change because one of its dependencies such as Kubernetes or scaler dependencies. However, these are the only exceptions to the rule.

### Custom Resource Definitions (CRDs)

Our custom resource definitions (CRDs) have a separate versioning scheme than the KEDA runtime given they have an `apiVersion`. For this, we follow the [official Kubernetes API versioning policy](https://kubernetes.io/docs/reference/using-api/#api-versioning).

KEDA is allowed to make breaking changes to the CRDs, when a new `apiVersion` is introduced.

> ### Experimental features

Author: @zroubalik I was feeling lucky and took a stab at it. Thoughts?

Reviewer 2: @tomkerkhove lookin good, thanks for tackling this!
